### ISSUE TITLE
Recognize blacklisted IP addrs stored in the file system namespace

### ISSUE THREAD

**Issue by krader1961** [pyed/ipfilter#29]
Issue #22 may be useful in some contexts but is inflexible. Do you really want to deal with reloading the file whenever it changes? It would be better to support a dynamic store such as a SQLite DB. My preference is to leverage the file system namespace. When I was running Apache+Wordpress and wanted to dynamically blacklist IP addrs with minimal overhead and without the need to introduce another dependency. So I put the following config statements in my *httpd.conf* file.

The basic idea is that blacklisted IP addresses are stored as nodes in the file system namespace under a *blacklisted* directory. To make it scale to thousands of entries I sharded the addresses into directories based on the IP addr prefix. For example, to blacklist `1.2.3.4` my tools would create a file named *blacklisted/1/2/1.2.3.4* whose content would contain text explaining when and why it was blacklisted. The explanatory text was optional. The address was blacklisted even if the file was empty.

While I chose to use a *blacklisted* directory in the Apache web server document root this should be configurable. It should support a path prefix relative to the Caddy document root and an absolute path prefix.

```
# Check if the IP address has been blacklisted by seeing if the address exists
# as a file in the "blacklisted" sub-tree of the web server document tree. Do
# not use REMOTE_ADDR as that is a HTTP header that can be spoofed. Note that
# using CONN_REMOTE_ADDR assumes no proxy or load balancer between this web
# server and the public network ingress router. If you have such an
# arrangement you'll need to replace CONN_REMOTE_ADDR with the name of the
# HTTP header you trust that identifies the originating IP address.
#
# This handles IPv4 and IPv6 addresses including "127.0.0.1" and "::1" (i.e.,
# "localhost"). This is useful for testing the rules contained herein as you
# can manually add a localhost address then issue a request and verify that
# the blacklist is effective.
<If "%{CONN_REMOTE_ADDR} =~ /^([0-9a-f]*)[.:]([0-9a-f]*)[.:]/ && \
    -f \"%{DOCUMENT_ROOT}/blacklisted/$1/$2/%{CONN_REMOTE_ADDR}\"">
    SetEnv BLACKLISTED "yes"
    RewriteRule ^ ${DOC_ROOT}/forbidden.php [END,E=error-notes:address-blacklisted]
</If>
<ElseIf "%{REMOTE_ADDR} =~ /^([0-9a-f]*)[.:]([0-9a-f]*)[.:]/ && \
    -f \"%{DOCUMENT_ROOT}/blacklisted/$1/$2/%{REMOTE_ADDR}\"">
    # We fallback to checking the REMOTE_ADDR value because if the source is being
    # proxied we still want to reject its traffic if it has already been
    # blacklisted.
    SetEnv BLACKLISTED "yes"
    RewriteRule ^ ${DOC_ROOT}/forbidden.php [END,E=error-notes:address-blacklisted]
</ElseIf>
<Else>
    SetEnv BLACKLISTED "no"
</Else>
```

**Comment by pyed** [comment]
neat solution, should be efficient and easy to implement, I've been busy for the last year and probably will be for another, haven't had the time to dive back in and write some code, your PR are welcome though.

Thanks for taking the time to propose a solution.

**Comment by krader1961** [comment]
> ...should be efficient and easy to implement...

Indeed. I took a few minutes to write a program in C that uses the `rdtsc` instruction to time how long `stat()` takes on my macOS server. If the directory is in the filesystem cache and the file doesn't exist (which will be the most common case) the syscall takes 5 microseconds. If the file does exist it's 38 microseconds. If the directory isn't in the filesystem cache and we have to perform I/O it's 2 milliseconds on my system. Which uses a SSD so may be faster than other peoples servers.

**Comment by krader1961** [comment]
Also, FWIW, this mechanism is meant to work with automated analysis of web server activity. When I was using a LAMP+Wordpress stack I ran a web server monitoring daemon I wrote to analyze the Apache log file for indicators that a system was attacking me. Or at least looking for vulnerabilities that could be exploited (e.g., the presence of WordPress plugins I had never installed). That monitoring daemon would create a file in the *blacklisted* directory when it detected requests from malware. I haven't yet ported that tool to work with my new Caddy web server but doing so should be trivial since it mostly relies on the access log format which is mostly compatible between the two environments.

**Comment by qdm12** [comment]
It would be nice to have:
1. the plugin generate the files as soon as an IP tries to access a bad website (such as public IP) or with a broken TLS version for example.
1. the plugin maintain a timestamp of all bad IPs and unblacklist them after a few days
1. refuse connections directly at the listener level of Caddy?

Just some ideas let me know what you think!

**Comment by krader1961** [comment]
@qdm12 I didn't write this plugin. I've only enhanced it. Nonetheless, I think the features you requested are outside the scope of this plugin and thus should not be implemented by this plugin. In fact, I think it is questionable whether any Caddy plugin should implement those features since they can all be done via an external daemon that monitors the Caddy log files. Anything that slows down the core behaviors of Caddy (e.g., returning content in response to HTTTP "GET" requests) needs a lot of justification. And the features you requested don't rise to that level.

**Comment by qdm12** [comment]
Regarding your enhancement, is it caching the blacklisted IP address from the persistent storage? Or is it accessing the persistent storage to check for each IP?

> Nonetheless, I think the features you requested are outside the scope of this plugin and thus should not be implemented by this plugin.

This plugin is an IP filter so it should filter IP addresses, that is true. But I think many users would appreciate the option to auto-generate/maintain blacklisting of IP addresses as the information is accessible from within Caddy (listener level). I'll write the plugin myself considering your answer.

> In fact, I think it is questionable whether any Caddy plugin should implement those features since they can all be done via an external daemon that monitors the Caddy log files.

Fail2ban can indeed be used at the firewall level which is best. But some environments don't have access to the firewall or don't have iptables, and it would be good to have a built-in mechanism to drop connections.

> Anything that slows down the core behaviors of Caddy (e.g., returning content in response to HTTTP "GET" requests) needs a lot of justification

Receiving a GET request from a bot and sending back status code 404 is as consuming as dropping the connection in my opinion.

> And the features you requested don't rise to that level.

All these were to be defined by user parameters. And see above.

**Comment by krader1961** [comment]
> Regarding your enhancement, is it caching the blacklisted IP address from the persistent storage? Or is it accessing the persistent storage to check for each IP?

The latter. Precisely because doing so is extremely inexpensive and thus fast enough for pretty much any web server. See my previous [comment](https://github.com/pyed/ipfilter/issues/29#issuecomment-440120811).

> But I think many users would appreciate the option to auto-generate/maintain blacklisting of IP addresses as the information is accessible from within Caddy (listener level).

The problem is that non-trivial filtering of the sort you want usually requires more context than a single HTTP request provides. For example, my web server log monitoring daemon considers whether a given host is reporting different user-agent strings in a short interval. It also considers whether a large number of requests for an invalid URL are received from multiple hosts in a short interval. Both of those situations are a very strong signal that the systems are making an attack.

Keeping all that context in a plugin like `ipfilter` is possible but can itself cause problems. Not least of which is slowing down the response to every single valid request. It is better to delegate such analysis to a tool running independent of the web server.

There are a few scenarios that might be considered within the scope of this plugin. For example, a host that generates a request in an invalid `Host:` header. But that is already handled efficiently by Caddy without needing to explicitly block the source. Assuming, of course, you have configured Caddy to reject any request that does not specify a hostname valid for your web server. And if you haven't done that then how exactly would this plugin be configured to blacklist the source? For example, my config begins with these lines:

```
localhost:80,
www.skepticism.us {
```

Any request that does not specify a HTTP `Host:` header using either of those names is automatically rejected without the need for this plugin. Which rejects 99% of malware requests to my web server since they almost always specify my IP address, domain name, or other invalid value for the `Host:` header.
